### ISSUE TITLE
Method naming style inconsistent with Swift book

### ISSUE THREAD

**Issue by obuseme** [raywenderlich/swift-style-guide#92]
From Apple's Swift Book:

“As in Objective-C, the name of a method in Swift typically refers to the method’s first parameter using a preposition such as with, for, or by, as seen in the incrementBy(_:) method from the preceding Counter class example. The use of a preposition enables the method to be read as a sentence when it is called.”

Excerpt From: Apple Inc. “The Swift Programming Language.” iBooks. https://itun.es/us/jEUH0.l

Thus, I think our method naming style is inconsistent such that `upvoteBy(amount: Int)` matches Apple's recommendation, but `combineWithString(incoming: String, options: Dictionary?)` should be `combineWith(incoming: String, options: Dictionary?)`

If ya'll okay with this change, I can submit a PR.

**Comment by gregheo** [comment]
Hey @obuseme, a pull request would be great!

`incrementBy` is a good example; `combineWith` seems less than ideal since we don't have a whole class and it's not clear what you're combining. But maybe that doesn't matter for a style guide.

**Comment by obuseme** [comment]
Sounds good, I'll get one prepped.

**Comment by obuseme** [comment]
Well the existing example in the style guide is `func upvoteBy(amount: Int)` so rather than add another example that is of the form `xyxBy(abc: Int)` I'll keep `combineWith(incoming: String, options: Dictionary?)`  IMHO I think the important info is conveyed - how the method should be named.

**Comment by obuseme** [comment]
PR #100

**Comment by jawwad** [comment]
Hmm, shouldn't this be the other way around? i.e. `upvoteByAmount(amount: Int)` should the the preferred naming style as the method name should refer to the first parameter by name for methods.

**Comment by obuseme** [comment]
@jawwad Check out page 160 in The Swift Programming Language, that's where this discrepacy caught my eye when having a debate at my day job about a method naming standard.   The example shown there to describe how methods should be named is:

    func​ ​incrementBy​(​amount​: ​Int​, ​numberOfTimes​: ​Int​)

I guess Apple's specific description is a little ambiguous.  See the quote in my original post here.

My preference is

    combineWith(incoming: String, options: Dictionary?)
    upvoteBy(amount: Int)

And I could be convinced to use:
    combineWithIncoming(incoming: String, options: Dictionary?)
    upvoteByAmount(amount: Int)

But don't think either of these are appropriate
    combineWithString(incoming: String, options: Dictionary?)
    upvoteByInt(amount: Int)

And `combineWithString` is what's in the guide currently.

**Comment by conradwt** [comment]
@obuseme The following seems a bit repetitive being the first parameter of the function matches the last word of the method name:

```swift
combineWithIncoming(incoming: String, options: Dictionary?)
upvoteByAmount(amount: Int)
```

Next, it's best to not use type information within method's name unless it's absolutely necessary to do it.  Thus, I prefer my functions to be intention revealing with embedding type information like the following:

```swift
func authenticate( username: String, password: String )
```

but the following could also be acceptable:

```swift
func authenticateWith( username: String, password: String )
```

Lastly, I would file bug report against the documentation if it's not clear to the reader or incomplete in regards to the context.  Swift is a new language without any strong function naming conventions.  Thus, one should be consistent with the naming across the codebase and do what works best for your organization.

**Comment by jawwad** [comment]
@obuseme, Hmm, interesting, hadn't noticed that in the guide.

So I believe the general style guideline is to refer to the first parameter in the method name, and the style guide also states:
"For methods, follow the standard Apple convention of referring to the first parameter in the method name"

But now the examples don't refer to the first parameter in the method name, so it actually doesn't match what the text above states.

If I were to guess I'd say that the incrementBy name doesn't have incrementByAmount since its clear from the context and its ok to deviate from the standard convention or guideline if it makes more sense for a better name in swift.

I think that `combineWithString` is still ok since the name doesn't exactly have match the parameter being used. For example if you take a look at the method `func stringByReplacingOccurrencesOfString(_ target: String, withString replacement: String` the internal parameter is named target but the method is not named `stringByReplacingOccurrencesOfTarget`. The name `combineWithIncoming` might be ok but it doesn't tell me what incoming is. If I myself had to include incoming in the method I might name it `combineWithIncomingString`. But I do agree that `upvoteByInt` would be weird so I would either go with `upvoteByAmount`  or `upvoteBy`.

Maybe we should change the examples? Or maybe we can say that in general the preferred method is to include the name, but its ok to deviate if it makes sense?

**Comment by obuseme** [comment]
@jawwad I don't think it's effective to debate this by example because there's many examples in either direction, eg on `UIAlertView` : `initWithTitle:(NSString *)title message:(NSString *)message...'  (there's no reference to "String" in the parameter names.

I'm advocating for consistency in our own standards.

I don't think it's appropriate in our tutorials to allow for deviations "if it makes sense."

Don't get me wrong, as experienced developers, there's totally cases where this is appropriate in the real world.

When writing tutorials for readers with a very wide range of experience, consistency in method naming will help the readers easily identify methods, how they are named, and connect the dots within the same tutorial, and even across different tutorials by different authors.

Additionally, being a little selfish here, it also helps the writers and tech editors easily identify right/wrong method names when writing/editing, without needing to go through a subjective opinion-based review of how the method is named.

Finally, with the lack of industry standard in Swift style, and with Apple even being ambiguous with their own advocacy for method names. this is an opportunity for us to be thought leaders in Swift style (Google search for "swift style guide" brings up this project as first result).

I'm just advocating for consistency, and right now, our guide isn't consistent.

To sum it up, what's everyone's opinion?  Methods named like
1) `combineWithString(incoming: String, options: Dictionary?)`
2) `combineWithIncoming(incoming: String, options: Dictionary?)`
3) `combineWith(incoming: String, options: Dictionary?)`

My vote is for 3, though I'd be okay with 2.

And no, I'm not talking about: http://en.wikipedia.org/wiki/Style_%28Taylor_Swift_song%29
:]

**Comment by hollance** [comment]
I think you don't need the *type* in the method name, as you already have the type in the parameter. (This isn't always true for Obj-C code, which may use `id`, a type that doesn't tell you anything.)

However, I think it's important to have the method name be explicit about what the first parameter *means* or represents.

**Comment by jawwad** [comment]
@obuseme: I think we've just proven the saying that "There are two hard things in computer science: cache invalidation, naming things, and off-by-one errors"! 

I'm ok with 2 as well, but like 1 better. I don't like 3. Maybe this example is too ambiguous and we should change it? Im definitely in favor of consistency though. You do bring up some good points. 

And you weren't talking about it but you still were thinking about it! :]

**Comment by jawwad** [comment]
It also depends on what class this method is in, hard to say what the best name would be without the context that its in. I think @hollance put it best, that the method name be explicit about what the first parameter means or represents.

**Comment by RobertGummesson** [comment]
Apple is not very consistent with their own preference. On the same page where they have they incrementBy() example, they also have:
```func advanceToLevel(level: Int)``` and not ```func advanceTo(level: Int)``` which would make it rather unclear.
Another example on the same page:
```func moveByX(deltaX: Double, y deltaY: Double)```

Surely no one would prefer ```func moveBy(deltaX: Double, y deltaY: Double)```?

I'm with @hollance, the method name should be explicit about what the first parameter means or represents.

**Comment by hollance** [comment]
`incrementBy()` makes sense, because there's no question of what you're incrementing by. 

Of course it will be a number, so writing `incrementByNumber()` is silly. And because the type of the number is known from the parameter, writing `incrementByInt()` is also superfluous.

But writing just `advanceTo()` is unclear. Advance to what? So yeah, there it makes perfect sense to say `advanceToLevel()`.

**Comment by RobertGummesson** [comment]
@hollance Are you suggesting the other example should be ```combineWith()``` then? 

I'm not convinced about ```combineWithIncoming(incoming: String, options: Dictionary?)```. An incoming what? In my mind ```combineWithString(incoming: String)```, ```combineWithColor(color: UIColor)``` etc would make it more obvious. 

What Apple methods or examples do we actually have that ends with With(), ex  ```methodNameWith(someParameter)```?

**Comment by hollance** [comment]
I'd have to know the context in which `combineWith...` is being used.

**Comment by pointum** [comment]
Instead of thinking about how a method declaration looks like, shouldn't it be more important to consider how a method call would be read like? Since the guide already requires using descriptive names for variables, then the first argument meaning should be clear without having to to put a label in the method name:

`someText.combineWith(otherText)` 
`counter.incrementBy(6)`
`counter.incrementBy(amount)`
`someColor.combineWith(redColor, alpha: 0.6)`
